enterprise/ha/src/main/java/org/neo4j/kernel/ha/cluster/SwitchToSlave.java

```diff
@@ -262,8 +263,9 @@ public URI switchToSlave( LifeSupport haCommunicationLife, URI me, URI masterUri
 
         monitor.switchToSlaveStarted();
 
-        // Wait for current transactions to stop first
-        long deadline = SYSTEM_CLOCK.currentTimeMillis() + config.get( HaSettings.state_switch_timeout );
+        // Wait a short while for current transactions to stop first, just to be nice.
+        // We can't wait forever since switching to our designated role is quite important.
+        long deadline = SYSTEM_CLOCK.currentTimeMillis() + config.get( HaSettings.internal_state_switch_timeout );
```

Comment:
I guess a fair question would be: why wait at all? What is it that these transactions will be able to do in that time that we are waiting for them?

It is pretty fair to assume that the master is gone here, so they cannot write. So the only transactions that _can_ finish are read-only transactions. Am I correct?

I'm wondering if it could be better to mark the transactions with some sort of failure mark that permits read operations for transactions that have not performed any write operations before (to allow read-only transactions to continue), but fails any write operations, or any operations at all on transactions that have already performed write operations.

With such a change we could perhaps completely remove the wait.

Granted, that change would be quite a lot bigger...